### PR TITLE
Feat S3 adapter

### PR DESCRIPTION
**Pull request by lohanidamodar**

### CODE REVIEW

File: src/Storage/Device/S3.php [hunk 1]
```diff
@@ -6,6 +6,30 @@
 
 class S3 extends Device
 {
+    protected $accessKey;
+    protected $secretKey;
+    protected $bucket;
+    protected $region;
+    protected $acl='private';
+    protected $root = 'temp';
+    private $headers = array(
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:15]: use newer `[]` syntax for PHP arrays

Comment (eldadfux) [reply src/Storage/Device/S3.php:15]: usually protected is better than private

File: src/Storage/Device/S3.php [hunk 2]
```diff
@@ -6,6 +6,30 @@
...
+		'Host' => '', 'Date' => '', 'Content-MD5' => '', 'Content-Type' => ''
+	);
+    private $response;
+    private $amzHeaders;
+
+    public function __construct($root='', $accessKey,$secretKey, $bucket, $region, $acl) {
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:21]: missing space after: `$accessKey`

Comment (eldadfux) [inline src/Storage/Device/S3.php:21]: default value for the first key don't do anything as you have to pass it anyway.

File: src/Storage/Device/S3.php [hunk 3]
```diff
@@ -6,6 +6,30 @@
...
+        $this->accessKey = $accessKey;
+        $this->secretKey = $secretKey;
+        $this->bucket = $bucket;
+        $this->region = $region;
+        $this->root = $root;
+        $this->acl = $acl;
+        $this->headers['Host'] = $this->bucket.'.s3.amazonaws.com';
+        $this->__resetResponse();
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:29]: Not common to use `_` for PHP private/protected methods.

File: src/Storage/Device/S3.php [hunk 4]
```diff
@@ -27,7 +51,7 @@ public function getDescription():string
      */
     public function getRoot():string
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:52]: I think we should have a space between `:` and `string`

File: src/Storage/Device/S3.php [hunk 5]
```diff
@@ -37,7 +61,13 @@ public function getRoot():string
      */
     public function getPath($filename):string
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:62]: missing type for $filename

File: src/Storage/Device/S3.php [hunk 6]
```diff
@@ -55,7 +85,7 @@ public function getPath($filename):string
      */
     public function upload($source, $path):bool
     {
-        return false;
+        return $this->write($path,$source);
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:88]: missing space between params

File: src/Storage/Device/S3.php [hunk 7]
```diff
@@ -55,7 +85,7 @@ public function getPath($filename):string
      */
     public function upload($source, $path):bool
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:86]: missing types for $source and $path

File: src/Storage/Device/S3.php [hunk 8]
```diff
@@ -67,7 +97,19 @@ public function upload($source, $path):bool
      */
     public function read(string $path):string
     {
-        return '';
+        $this->__resetResponse();
+        $verb = 'GET';
+        $uri = $path;
+        $uri = $uri !== '' ? '/'.str_replace('%2F', '/', rawurlencode($uri)) : '/';
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:103]: Might be more readable: `$uri !== '' ? '/'.str_replace('%2F', '/', rawurlencode($uri)) : '/';` - why not parse $path directly?

might be better to use: https://www.php.net/manual/en/function.urlencode.php?

File: src/Storage/Device/S3.php [hunk 9]
```diff
@@ -67,7 +97,19 @@ public function upload($source, $path):bool
...
+		$this->__getResponse($uri, $verb, false);
+
+		if ($this->response->error === false && $this->response->code !== 200)
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:106]: Indentation is off. Better use exceptions to handle errors here.

File: src/Storage/Device/S3.php [hunk 10]
```diff
@@ -78,11 +120,46 @@ public function read(string $path):string
      *
      * @return bool
      */
-    public function write(string $path, string $data):bool
+    public function write(string $path, string $data, string $contentType = 'text/plain'):bool
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:123]: Would be better to use empty string as default contentType value

File: src/Storage/Device/S3.php [hunk 11]
```diff
@@ -78,11 +120,46 @@ public function read(string $path):string
...
     {
-        return false;
+        $this->__resetResponse();
+        $uri = $path;
+        $metaHeaders = array();
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:127]: Use new array syntax: `[]`

File: src/Storage/Device/S3.php [hunk 12]
```diff
@@ -78,11 +120,46 @@ public function read(string $path):string
...
+			'data' => $data, 'size' => strlen($data),
+			'md5sum' => base64_encode(md5($data, true)),
+			'sha256sum' => hash('sha256', $data)
+		);
+        $input['type'] = $contentType;
+        if(isset($input['size']) && $input['size'] >= 0) {
+            $this->headers['Content-Type'] = $input['type'];
+            if (isset($input['md5sum'])) $this->headers['Content-MD5'] = $input['md5sum'];
+
+			if (isset($input['sha256sum'])) $this->amzHeaders['x-amz-content-sha256'] = $input['sha256sum'];
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:145]: using brackets is more common for the team, makes code more readable

File: src/Storage/Device/S3.php [hunk 13]
```diff
@@ -107,9 +184,30 @@ public function move(string $source, string $target):bool
...
      */
+
+     /* 
+     DELETE /my-second-image.jpg HTTP/1.1
+    Host: <bucket>.s3.<Region>.amazonaws.com
+    Date: Wed, 12 Oct 2009 17:50:00 GMT
+    Authorization: authorization string
+    Content-Type: text/plain    
+     */
+
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:195]: Remove un-required comments or add them as part of the docblock

File: src/Storage/Device/S3.php [hunk 14]
```diff
@@ -107,9 +184,30 @@ public function move(string $source, string $target):bool
...
     public function delete(string $path, bool $recursive = false):bool
     {
-        return false;
+        $this->__resetResponse();
+        $verb = 'DELETE';
+        $uri = $path;
+        $uri = $uri !== '' ? '/'.str_replace('%2F', '/', rawurlencode($uri)) : '/';
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:201]: Seems like some part of handling the request are repetitive, should we abstract it with a protected `call` method?

File: src/Storage/Device/S3.php [hunk 15]
```diff
@@ -123,7 +221,8 @@ public function delete(string $path, bool $recursive = false):bool
      */
     public function getFileSize(string $path):int
     {
-        return 0;
+        $res = $this->__getInfo($path);
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:224]: Avoid `_` as a naming convention.

File: src/Storage/Device/S3.php [hunk 16]
```diff
@@ -137,7 +236,8 @@ public function getFileSize(string $path):int
      */
     public function getFileMimeType(string $path):string
     {
-        return '';
+        $res = $this->__getInfo($path);
+        return $res['type'];
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:240]: validate `type` exists.

File: src/Storage/Device/S3.php [hunk 17]
```diff
@@ -151,7 +251,8 @@ public function getFileMimeType(string $path):string
      */
     public function getFileHash(string $path):string
     {
-        return '';
+        $res = $this->__getInfo($path);
+        return $res['hash'];
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:255]: validate `hash` exists.

File: src/Storage/Device/S3.php [hunk 18]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
     {
         return 0.0;
     }
+
+
+    private function __resetResponse()
+    {
+        $this->response = new \stdClass;
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:301]: Indentation is off. Handle error consistently using exceptions.

File: src/Storage/Device/S3.php [hunk 19]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
...
+		$this->response->body = null;
+		$this->response->headers = array();
+    }
+
+    private function __getInfo(string $path) {
+        $this->__resetResponse();
+        $verb = 'HEAD';
+        $uri = $path;
+        $uri = $uri !== '' ? '/'.str_replace('%2F', '/', rawurlencode($uri)) : '/';
+		$rest = $this->__getResponse($uri,$verb,false);
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:312]: Indentation is off. Use brackets for conditions.

File: src/Storage/Device/S3.php [hunk 20]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
...
+
+    /**
+	* Generate the headers for AWS Signature V4
+	* @param string $method
+	* @param string $uri
+	* @return string
+	*/
+    private function __getSignatureV4($method, $uri):string
+	{	
+        $parameters = array();
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:330]: Indentation is off. Use brackets for `if`s and `for`s.

Comment (eldadfux) [inline src/Storage/Device/S3.php:330]: newer php array syntax: `[]`

File: src/Storage/Device/S3.php [hunk 21]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
...
+		return $rest->code == 200 ? $rest->headers : false;
+    }
+
+    /**
+	* Generate the headers for AWS Signature V4
+	* @param string $method
+	* @param string $uri
+	* @return string
+	*/
+    private function __getSignatureV4($method, $uri):string
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:328]: no `_`

Comment (eldadfux) [inline src/Storage/Device/S3.php:328]: missing types for params

File: src/Storage/Device/S3.php [hunk 22]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
...
+	* @return object | false
+	*/
+	private function __getResponse(string $uri,  string $verb = 'GET',$data='')
+	{
+		$url = 'https://' . ($this->headers['Host'] !== '' ? $this->headers['Host'] : $this->endpoint) . $uri;
+
+		// Basic setup
+		$curl = curl_init();
+		curl_setopt($curl, CURLOPT_USERAGENT, 'utopia-php/storage');
+
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:407]: Multiple extra spaces

File: src/Storage/Device/S3.php [hunk 23]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
...
+		
+
+		curl_setopt($curl, CURLOPT_URL, $url);
+
+		
+		// Headers
+		$httpHeaders = array(); 
+		$this->amzHeaders['x-amz-date'] = gmdate('Ymd\THis\Z');
+
+		if (!isset($this->amzHeaders['x-amz-content-sha256']))
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:417]: brackets

File: src/Storage/Device/S3.php [hunk 24]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
...
+		curl_setopt($curl, CURLOPT_HTTPHEADER, $httpHeaders);
+		curl_setopt($curl, CURLOPT_HEADER, false);
+		curl_setopt($curl, CURLOPT_RETURNTRANSFER, false);
+		curl_setopt($curl, CURLOPT_WRITEFUNCTION, array(&$this, '__responseWriteCallback'));
+		curl_setopt($curl, CURLOPT_HEADERFUNCTION, array(&$this, '__responseHeaderCallback'));
+		curl_setopt($curl, CURLOPT_FOLLOWLOCATION, true);
+
+		// Request types
+		switch ($verb)
+		{
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:441]: Indentations is off.

File: src/Storage/Device/S3.php [hunk 25]
```diff
@@ -193,4 +294,265 @@ public function getPartitionTotalSpace():float
...
+		if ($strlen <= 2) return $strlen;
+		if (substr($data, 0, 4) == 'HTTP')
+			$this->response->code = (int)substr($data, 9, 3);
+		else
+		{
+			$data = trim($data);
+			if (strpos($data, ': ') === false) return $strlen;
+			list($header, $value) = explode(': ', $data, 2);
+			$header = strtolower($header);
+			if ($header == 'last-modified')
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:543]: Probably better use a switch case here.

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/Storage/Device/S3.php (hunk 7 above)
Comment (lohanidamodar) [reply src/Storage/Device/S3.php]: This is because the `Device` has as such, if i mention type here, I get an error

Comment (eldadfux) [reply src/Storage/Device/S3.php]: You can fix this at the device level as well 👍🏻

File: src/Storage/Device/S3.php [hunk 26]
```diff
@@ -121,9 +214,10 @@ public function delete(string $path, bool $recursive = false):bool
      *
      * @return int
      */
-    public function getFileSize(string $path):int
+    public function getFileSize(string $path): int
     {
-        return 0;
+        $res = $this->getInfo($path);
+        return $res['size'];
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:220]: It's a best practice to validate the key actually exist. Something like: `return (isset($res['size'])) ? $res['size'] : 0;` or the shorter syntax: `return $res['size'] ?? 0;`

File: src/Storage/Device/S3.php [hunk 27]
```diff
@@ -189,8 +285,295 @@ public function getPartitionFreeSpace():float
...
+
+    private function getInfo(string $path)
+    {
+        $this->resetResponse();
+        $verb = 'HEAD';
+        $uri = $path;
+        $uri = $uri !== '' ? '/' . str_replace('%2F', '/', rawurlencode($uri)) : '/';
+        $rest = $this->getResponse($uri, $verb, false);
+        if ($rest->error === false && ($rest->code !== 200 && $rest->code !== 404)) {
+            $rest->error = array('code' => $rest->code, 'message' => 'Unexpected HTTP status');
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:309]: avoid old array syntax. array() -> []

File: src/Storage/Device/S3.php [hunk 28]
```diff
@@ -189,8 +285,295 @@ public function getPartitionFreeSpace():float
...
+     *
+     * @param resource &$curl CURL resource
+     * @param string &$data Data
+     * @return integer
+     */
+    private function responseWriteCallback(&$curl, &$data)
+    {
+        $this->response->body .= $data;
+        return strlen($data);
+    }
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:529]: This all function can be just be passed as a callback on line 437, no real need for it to be a method.

File: src/Storage/Device/S3.php [hunk 29]
```diff
@@ -189,8 +285,295 @@ public function getPartitionFreeSpace():float
...
+
+    /**
+     * CURL header callback
+     *
+     * @param resource $curl CURL resource
+     * @param string $data Data
+     * @return integer
+     */
+    private function responseHeaderCallback($curl, $data)
```
Comment (eldadfux) [inline src/Storage/Device/S3.php:538]: Please double check if we can replace this method with something like this: https://github.com/appwrite/appwrite/blob/0.7.x/tests/e2e/Client.php#L190-L201 and just pass it as a callback on line 438

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/Storage/Device/S3.php (hunk 29 above)
Comment (lohanidamodar) [reply src/Storage/Device/S3.php]: I think we do need the header properly formatting for date and time headers. Passed it as a callback

**Comment by lohanidamodar** [comment]
@eldadfux We don't have tests for Upload and Move methods, I have tested them locally manually with a script, seems fine.
Still issue with how to run tests on CI, do I push test without credentials (This will fail the tests of course)?

**Comment by eldadfux** [comment]
> @eldadfux We don't have tests for Upload and Move methods, I have tested them locally manually with a script, seems fine.
> Still issue with how to run tests on CI, do I push test without credentials (This will fail the tests of course)?

Yes. Push the tests without credentials. In the tests take the credentials from env variables and I'll make sure to pass them with Travis later.

File: src/Storage/Device/S3.php (hunk 29 above)
Comment (eldadfux) [reply src/Storage/Device/S3.php]: Just seems like a bit of an overkill, do we use this headers anywhere in the code?

**Review by eldadfux — APPROVED** [review]

**Pull request merged**